pkg/k8s/resource/diffstore.go

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package resource
+
+import (
+	"context"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/stream"
+
+	k8sRuntime "k8s.io/apimachinery/pkg/runtime"
+)
+
+// DiffStoreFactory creates diffStores. All stores from the same factory share the same underlying store, but each
+// diff store tracks the diff since the last time it was checked.
+type DiffStoreFactory[T k8sRuntime.Object] interface {
+	// NewStore creates a new store from this factory, all DiffStores share the same underlying store but each
+	// records its own diff list.
+	NewStore() DiffStore[T]
+
+	// Observe returns a channel which will be signalled every time there is at least one pending change.
+	// Multiple changes are coalesced. Calling `done` will drain and close the channel, it should be called when no
+	// longer using `onDiff` to free resources.
+	Observe() (onDiff <-chan struct{}, done func())
+}
+
+var _ DiffStoreFactory[*k8sRuntime.Unknown] = (*diffStoreFactory[*k8sRuntime.Unknown])(nil)
+
+type diffStoreParams[T k8sRuntime.Object] struct {
+	cell.In
+
+	LS         hive.Lifecycle
```

Comment:
Should be `LC`, but please use `Lifecycle` instead. Nicer to read when it's consistent with the other names.